File: skills/agent-swarm/scripts/backends/acp/permissions.py
"""Headless ACP permission policies; approval automation is not an OS sandbox."""

import json
import os
import re
import shlex


_ACTION_TYPE = re.compile(r"^[a-z][a-z0-9_]*$")
_ENV_RUNTIME_ENTRYPOINT = "$AGENT_SWARM_SKILL_DIR/scripts/agent_orchestrator.py"
_TRUSTED_SHELLS = frozenset(
    os.path.realpath(path)
    for path in (
        "/bin/bash",
        "/bin/sh",
        "/bin/zsh",
        "/usr/bin/bash",
        "/usr/bin/sh",
        "/usr/bin/zsh",
    )
    if os.path.exists(path)
)


def _kind(option):
    return str(option.get("kind") or option.get("name") or "").lower()


def _option(options, allow, *, once_only=False):
    for option in options:
        if not isinstance(option, dict) or not option.get("optionId"):
            continue
        kind = _kind(option)
        is_deny = any(token in kind for token in ("deny", "reject", "cancel"))
        is_allow = "allow" in kind or "approve" in kind
        if allow and once_only and "once" not in kind:
            continue
        if (allow and is_allow and not is_deny) or (not allow and is_deny):
            return option["optionId"]
    return None


def selected_option_allows(params, option_id):
    """Classify a decision by the offered option, never by opaque optionId text."""
    options = params.get("options") if isinstance(params, dict) else None
    for option in options if isinstance(options, list) else []:
        if not isinstance(option, dict) or option.get("optionId") != option_id:
            continue
        kind = _kind(option)
        denied = any(token in kind for token in ("deny", "reject", "cancel"))
        return not denied and any(token in kind for token in ("allow", "approve"))
    return False


def _locations(params):
    tool_call = params.get("toolCall") if isinstance(params, dict) else None
    candidates = []
    for owner in (params, tool_call):
        if not isinstance(owner, dict):
            continue
        value = owner.get("locations")
        if isinstance(value, list):
            candidates.extend(value)
    paths = []
    for location in candidates:
        path = location.get("path") if isinstance(location, dict) else location
        if not isinstance(path, str) or not os.path.isabs(path):
            return None
        paths.append(os.path.realpath(path))
    return paths or None


def _locations_declared(params):
    tool_call = params.get("toolCall") if isinstance(params, dict) else None
    for owner in (params, tool_call):
        if not isinstance(owner, dict) or "locations" not in owner:
            continue
        if owner.get("locations") not in (None, []):
            return True
    return False


def _inside(path, roots):
    for root in roots:
        try:
            if os.path.commonpath([path, root]) == root:
                return True
        except ValueError:
            continue
    return False


def _runtime_entrypoint_matches(value, runtime_entrypoint):
    if value == _ENV_RUNTIME_ENTRYPOINT:
        return True
    if not runtime_entrypoint or not os.path.isabs(value):
        return False
    return os.path.realpath(value) == os.path.realpath(runtime_entrypoint)


def _shell_tokens(script):
    if (
        not isinstance(script, str)
        or not script
        or "`" in script
        or "\n" in script
        or "\r" in script
    ):
        return None
    try:
        lexer = shlex.shlex(script, posix=True, punctuation_chars="|&;()<>")
        lexer.whitespace_split = True
        lexer.commenters = ""
        return list(lexer)
    except ValueError:
        return None


def _runtime_cli_request(params, *, cwd, runtime_entrypoint):
    """Recognize one narrowly-scoped Runtime CLI invocation from Codex ACP.

    Codex requests approval because the authenticated Runtime CLI must update
    its state outside the child workspace. Only the documented bootstrap,
    schema lookup, and JSON-to-Action forms are eligible. Shell chaining,
    redirection, alternate producers, and unrelated Runtime subcommands fail
    closed.
    """
    if not runtime_entrypoint:
        return False
    tool_call = params.get("toolCall") if isinstance(params, dict) else None
    if not isinstance(tool_call, dict) or tool_call.get("kind") != "execute":
        return False
    raw_input = tool_call.get("rawInput")
    if not isinstance(raw_input, dict):
        return False
    request_cwd = raw_input.get("cwd")
    if not isinstance(request_cwd, str) or not os.path.isabs(request_cwd):
        return False
    workspace = os.path.realpath(cwd)
    if not _inside(os.path.realpath(request_cwd), [workspace]):
        return False
    command = raw_input.get("command")
    if (
        not isinstance(command, list)
        or len(command) != 3
        or not all(isinstance(item, str) for item in command)
        or not os.path.isabs(command[0])
        or os.path.realpath(command[0]) not in _TRUSTED_SHELLS
        or command[1] not in {"-c", "-lc"}
    ):
        return False
    tokens = _shell_tokens(command[2])
    if not tokens:
        return False

    def is_runtime_command(parts):
        return (
            len(parts) >= 3
            and parts[0] == "python3"
            and _runtime_entrypoint_matches(parts[1], runtime_entrypoint)
        )

    if is_runtime_command(tokens):
        if tokens[2:] == ["bootstrap-cwd"]:
            return True
        return bool(
            len(tokens) == 4
            and tokens[2] == "action-schema"
            and _ACTION_TYPE.fullmatch(tokens[3])
        )

    prefix = "printf '%s' '"
    separator = "' | "
    if not command[2].startswith(prefix):
        return False
    remainder = command[2][len(prefix) :]
    if remainder.count(separator) != 1:
        return False
    encoded_payload, consumer_script = remainder.split(separator)
    if "'" in encoded_payload:
        return False
    consumer = _shell_tokens(consumer_script)
    if not consumer:
        return False
    try:
        payload = json.loads(encoded_payload)
    except (TypeError, ValueError):
        return False
    return bool(
        isinstance(payload, dict)
        and is_runtime_command(consumer)
        and len(consumer) == 6
        and consumer[2:4] == ["action", "--type"]
        and _ACTION_TYPE.fullmatch(consumer[4])
        and consumer[5:] == ["--stdin"]
    )


def decide_permission(
    params,
    *,
    policy,
    cwd,
    additional_directories=None,
    runtime_entrypoint=None,
):
    options = params.get("options") if isinstance(params, dict) else None
    options = options if isinstance(options, list) else []
    runtime_exception = False
    if policy == "prompt":
        raise RuntimeError("ACP permission policy 'prompt' has no headless UI")
    if policy == "allow_all":
        allow = True
    elif policy == "deny_all":
        allow = False
    elif policy == "allow_in_workspace":
        locations = _locations(params)
        roots = [os.path.realpath(cwd)]
        roots.extend(os.path.realpath(path) for path in (additional_directories or []))
        if locations:
            allow = all(_inside(path, roots) for path in locations)
        elif _locations_declared(params):
            allow = False
        else:
            runtime_exception = _runtime_cli_request(
                params,
                cwd=cwd,
                runtime_entrypoint=runtime_entrypoint,
            )
            allow = runtime_exception
    else:
        raise RuntimeError("unknown ACP permission policy: %s" % policy)
    option_id = _option(options, allow, once_only=runtime_exception)
    if option_id is None and allow:
        option_id = _option(options, False)
    if option_id is None:
        return {"outcome": {"outcome": "cancelled"}}
    return {"outcome": {"outcome": "selected", "optionId": option_id}}

File: skills/agents-orchestrator/scripts/backends/acp/permissions.ts
import { existsSync, realpathSync } from "node:fs";
import { isAbsolute, relative, resolve } from "node:path";

import type { RequestPermissionRequest } from "@agentclientprotocol/sdk";

import { RuntimeError, type RuntimeRecord } from "../../runtime_types.ts";

const ACTION_TYPE = /^[a-z][a-z0-9_]*$/u;
const ENV_RUNTIME_ENTRYPOINTS = new Set([
  "$AGENTS_ORCHESTRATOR_SKILL_DIR/scripts/agent_orchestrator.ts",
  "$AGENTS_ORCHESTRATOR_SKILL_DIR/scripts/bootstrap.ts",
]);
const TRUSTED_SHELLS = new Set(
  ["/bin/bash", "/bin/sh", "/bin/zsh", "/usr/bin/bash", "/usr/bin/sh", "/usr/bin/zsh"]
    .filter(existsSync)
    .map((path) => realpathSync(path)),
);

export interface PermissionDecision {
  selectedOptionId: string | null;
  allowed: boolean;
}

function payload(value: unknown): RuntimeRecord {
  if (value !== null && typeof value === "object" && !Array.isArray(value)) return value as RuntimeRecord;
  throw new TypeError("ACP permission request must be an official schema object");
}

function kind(option: RuntimeRecord): string {
  return String(option.kind ?? option.name ?? "").toLowerCase();
}

function option(options: unknown[], allow: boolean, onceOnly = false): string | null {
  for (const candidate of options) {
    if (candidate === null || typeof candidate !== "object" || Array.isArray(candidate)) continue;
    const item = candidate as RuntimeRecord;
    if (typeof item.optionId !== "string" || !item.optionId) continue;
    const value = kind(item);
    const deny = ["deny", "reject", "cancel"].some((token) => value.includes(token));
    const approve = value.includes("allow") || value.includes("approve");
    if (allow && onceOnly && !value.includes("once")) continue;
    if ((allow && approve && !deny) || (!allow && deny)) return item.optionId;
  }
  return null;
}

export function selectedOptionAllows(params: unknown, optionId: string): boolean {
  const options = payload(params).options;
  for (const candidate of Array.isArray(options) ? options : []) {
    if (candidate === null || typeof candidate !== "object" || Array.isArray(candidate)) continue;
    const item = candidate as RuntimeRecord;
    if (item.optionId !== optionId) continue;
    const value = kind(item);
    const denied = ["deny", "reject", "cancel"].some((token) => value.includes(token));
    return !denied && (value.includes("allow") || value.includes("approve"));
  }
  return false;
}

function canonical(path: string): string {
  try { return realpathSync(path); } catch { return resolve(path); }
}

function inside(path: string, roots: string[]): boolean {
  return roots.some((root) => {
    const segment = relative(root, path);
    return segment === "" || (!segment.startsWith("..") && !isAbsolute(segment));
  });
}

function locations(params: RuntimeRecord): string[] | null {
  const toolCall = params.toolCall;
  const candidates: unknown[] = [];
  for (const owner of [params, toolCall]) {
    if (owner === null || typeof owner !== "object" || Array.isArray(owner)) continue;
    const value = (owner as RuntimeRecord).locations;
    if (Array.isArray(value)) candidates.push(...value);
  }
  const paths: string[] = [];
  for (const location of candidates) {
    const path = location !== null && typeof location === "object" && !Array.isArray(location)
      ? (location as RuntimeRecord).path : location;
    if (typeof path !== "string" || !isAbsolute(path)) return null;
    paths.push(canonical(path));
  }
  return paths.length ? paths : null;
}

function locationsDeclared(params: RuntimeRecord): boolean {
  for (const owner of [params, params.toolCall]) {
    if (owner === null || typeof owner !== "object" || Array.isArray(owner)) continue;
    const item = owner as RuntimeRecord;
    if (!("locations" in item)) continue;
    if (item.locations !== null && !(Array.isArray(item.locations) && item.locations.length === 0)) return true;
  }
  return false;
}

function runtimeEntrypointMatches(value: string, runtimeEntrypoint?: string): boolean {
  if (ENV_RUNTIME_ENTRYPOINTS.has(value)) return true;
  if (!runtimeEntrypoint || !isAbsolute(value)) return false;
  if (canonical(value) === canonical(runtimeEntrypoint)) return true;
  const bootstrap = runtimeEntrypoint.replace(/agent_orchestrator\.ts$/u, "bootstrap.ts");
  return bootstrap !== runtimeEntrypoint && canonical(value) === canonical(bootstrap);
}

// Deliberately tiny POSIX lexer. It accepts words and the punctuation needed by
// the documented printf pipeline, while rejecting substitutions and malformed quotes.
function shellTokens(script: unknown): string[] | null {
  if (typeof script !== "string" || !script || /[`\n\r]/u.test(script)) return null;
  const tokens: string[] = [];
  let current = "";
  let quote: "'" | '"' | null = null;
  let started = false;
  const punctuation = new Set(["|", "&", ";", "(", ")", "<", ">"]);
  for (let index = 0; index < script.length; index += 1) {
    const char = script[index]!;
    if (quote !== "'" && char === "\\") {
      const next = script[index + 1];
      if (next === undefined) return null;
      current += next;
      started = true;
      index += 1;
      continue;
    }
    if (char === "'" || char === '"') {
      if (quote === null) { quote = char; started = true; continue; }
      if (quote === char) { quote = null; continue; }
      current += char;
      started = true;
      continue;
    }
    if (quote === null && /\s/u.test(char)) {
      if (started) { tokens.push(current); current = ""; started = false; }
      continue;
    }
    if (quote === null && punctuation.has(char)) {
      if (started) { tokens.push(current); current = ""; started = false; }
      tokens.push(char);
      continue;
    }
    if (quote !== "'" && char === "$" && script[index + 1] === "(") return null;
    current += char;
    started = true;
  }
  if (quote !== null) return null;
  if (started) tokens.push(current);
  return tokens;
}

function runtimeCliRequest(params: RuntimeRecord, cwd: string, runtimeEntrypoint?: string): boolean {
  if (!runtimeEntrypoint) return false;
  const toolCall = params.toolCall;
  if (toolCall === null || typeof toolCall !== "object" || Array.isArray(toolCall)) return false;
  const call = toolCall as RuntimeRecord;
  if (call.kind !== "execute") return false;
  const raw = call.rawInput;
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return false;
  const input = raw as RuntimeRecord;
  const requestCwd = input.cwd;
  if (typeof requestCwd !== "string" || !isAbsolute(requestCwd) || !inside(canonical(requestCwd), [canonical(cwd)])) return false;
  const command = input.command;
  if (!Array.isArray(command) || command.length !== 3 || !command.every((item) => typeof item === "string")) return false;
  const [shell, flag, script] = command as string[];
  if (!isAbsolute(shell!) || !TRUSTED_SHELLS.has(canonical(shell!)) || !new Set(["-c", "-lc"]).has(flag!)) return false;
  const tokens = shellTokens(script);
  if (!tokens) return false;
  const runtimeCommand = (parts: string[]): boolean => (
    parts.length >= 3 &&
    (parts[0] === "bun" || (isAbsolute(parts[0]!) && canonical(parts[0]!) === canonical(process.execPath))) &&
    runtimeEntrypointMatches(parts[1]!, runtimeEntrypoint)
  );
  if (runtimeCommand(tokens)) {
    if (tokens.length === 3 && tokens[2] === "bootstrap-cwd") return true;
    return tokens.length === 4 && tokens[2] === "action-schema" && ACTION_TYPE.test(tokens[3]!);
  }
  const prefix = "printf '%s' '";
  const separator = "' | ";
  if (!script!.startsWith(prefix)) return false;
  const remainder = script!.slice(prefix.length);
  if (remainder.split(separator).length !== 2) return false;
  const [encoded, consumerScript] = remainder.split(separator) as [string, string];
  if (encoded.includes("'")) return false;
  let decoded: unknown;
  try { decoded = JSON.parse(encoded); } catch { return false; }
  const consumer = shellTokens(consumerScript);
  return Boolean(
    decoded !== null && typeof decoded === "object" && !Array.isArray(decoded) && consumer &&
    runtimeCommand(consumer) && consumer.length === 6 && consumer[2] === "action" &&
    consumer[3] === "--type" && ACTION_TYPE.test(consumer[4]!) && consumer[5] === "--stdin",
  );
}

export function decidePermission(
  request: RequestPermissionRequest | RuntimeRecord,
  options: {
    policy: string;
    cwd: string;
    additionalDirectories?: string[];
    runtimeEntrypoint?: string;
  },
): PermissionDecision {
  const params = payload(request);
  const offered = Array.isArray(params.options) ? params.options : [];
  let allow: boolean;
  let runtimeException = false;
  if (options.policy === "prompt") throw new RuntimeError("ACP permission policy 'prompt' has no headless UI");
  if (options.policy === "allow_all") allow = true;
  else if (options.policy === "deny_all") allow = false;
  else if (options.policy === "allow_in_workspace") {
    const paths = locations(params);
    const roots = [canonical(options.cwd), ...(options.additionalDirectories ?? []).map(canonical)];
    if (paths) allow = paths.every((path) => inside(path, roots));
    else if (locationsDeclared(params)) allow = false;
    else {
      runtimeException = runtimeCliRequest(params, options.cwd, options.runtimeEntrypoint);
      allow = runtimeException;
    }
  } else throw new RuntimeError(`unknown ACP permission policy: ${options.policy}`);
  let selectedOptionId = option(offered, allow, runtimeException);
  if (selectedOptionId === null && allow) selectedOptionId = option(offered, false);
  if (selectedOptionId === null) return { selectedOptionId: null, allowed: false };
  return { selectedOptionId, allowed: selectedOptionAllows(params, selectedOptionId) };
}
